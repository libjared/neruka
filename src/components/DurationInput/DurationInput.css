.DurationInput-main {
  display: inline-block;
  user-select: none;
}
.DurationInput-main:focus {
  outline: none;
}

.DurationInput-area {
  border-bottom: 2px solid #3c4043;
  cursor: pointer;
}
.DurationInput-area.editing {
  border-bottom: 2px solid #8ab4f8;
  cursor: default;
}

.Digit, .Label {
  color: #5f6368;
}
.Digit.lit, .Label.lit {
  color: #e8eaed;
}

.DurationInput-area .Label.seconds {
  border-left: 1px solid transparent;
}
.DurationInput-area.editing .Label.seconds {
  border-left: 1px solid white;
}
