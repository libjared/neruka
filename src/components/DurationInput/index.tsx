import classNames from 'classnames';
import { useState } from 'react';
import { Digit, Label } from '../Digit';
import './DurationInput.css';

function renderDigits(hhmmss: number[], lightLength: number | null): JSX.Element[] {
  const elements: JSX.Element[] = [];
  const lightLen = lightLength ?? 9999;
  for (let i = 0; i < hhmmss.length; i++) {
    const alignedIdx = i + (6 - hhmmss.length);
    const isLit = 6 - i <= lightLen;
    elements.push(<Digit key={alignedIdx} value={hhmmss[i]} isLit={isLit} />);
    if (alignedIdx === 1) {
      elements.push(<Label key="h" label="h" isLit={isLit} />);
    }
    if (alignedIdx === 3) {
      elements.push(<Label key="m" label="m" isLit={isLit} />);
    }
    if (alignedIdx === 5) {
      elements.push(<Label key="s" label="s" isLit={isLit} />);
    }
  }
  return elements;
}

// returns a new left-padded array of digits.
function digitsFromDuration(duration: Duration): number[] {
  return [
    Math.floor((duration.hours || 0) / 10),
    Math.floor((duration.hours || 0) % 10),
    Math.floor((duration.minutes || 0) / 10),
    Math.floor((duration.minutes || 0) % 10),
    Math.floor((duration.seconds || 0) / 10),
    Math.floor((duration.seconds || 0) % 10),
  ];
}

// remove left-padded zeroes from an array of 6 digits.
// will always leave at least one element, even if it's also a zero.
// returns a new array.
function trimDigits(digits: number[]): number[] {
  if (digits.length !== 6) throw new Error('Expected digits to have a length of 6.');

  const newDigits = [ ...digits ];
  for (let i = 0; i < newDigits.length - 1; i++) {
    if (newDigits[i] === 0) {
      newDigits.shift();
      i--;
    } else {
      break;
    }
  }
  return newDigits;
}

// prepend zeroes until the array has a length of 6.
// returns a new array.
function padDigits(digits: number[]) : number[] {
  if (digits.length > 6) throw new Error('Expected digits to be fewer or equal to 6.');

  return [ ...(new Array(6).fill(0)), ...digits ].reverse().slice(0, 6).reverse();
}

type DurationInputProps = {
  duration: Duration,
  onFinishEditing: (newDuration: Duration, immediatelyStart: boolean) => void
};

function DurationInput({ duration, onFinishEditing }: DurationInputProps) {
  const [ isEditing, setEditing ] = useState<boolean>(false);

  // hhmmss, right-aligned, length is [0,6], no left-padding with zeroes.
  // represents the in-progress inputting of a duration.
  // only valid when isEditing true; otherwise null.
  const [ wipDigits, setWipDigits ] = useState<number[] | null>(null);

  // how many wipDigits from the right should be lit up?
  // the h/m/s labels are also lit, but implicitly (ie they don't contribute to the count).
  // only valid when isEditing true, otherwise null.
  const [ lightLength, setLightLength ] = useState<number | null>(null);

  const onFocus: React.FocusEventHandler<HTMLDivElement> = () => {
    if (isEditing) throw new Error('Expected isEditing to be false.');
    if (wipDigits !== null) throw new Error('Expected wipDigits to be null.');

    // on edit, every digit is present, but unlit, like a placeholder
    setEditing(true);
    setWipDigits(digitsFromDuration(duration));
    setLightLength(0);
  };

  const saveAndQuit = (startImmediately: boolean) => {
    if (!isEditing) throw new Error('Expected isEditing to be true.');
    if (wipDigits === null) throw new Error('Expected wipDigits to be non-null.');

    const hhmmss = padDigits(wipDigits);
    const newDuration: Duration = {
      hours: hhmmss[0] * 10 + hhmmss[1],
      minutes: hhmmss[2] * 10 + hhmmss[3],
      seconds: hhmmss[4] * 10 + hhmmss[5]
    };

    // TODO: run digits through a duration normalizer
    // 99m => 1h39m
    // 0m10s => 10s
    // 01s => 1s
    // wow date-fns is half-baked. I gotta normalize durations myself?
    setEditing(false);
    setWipDigits(null);
    setLightLength(null);
    onFinishEditing(newDuration, startImmediately);
  };

  const onBlur: React.FocusEventHandler<HTMLDivElement> = () => {
    saveAndQuit(false);
  };

  const onKeyDown: React.KeyboardEventHandler<HTMLDivElement> = (ev) => {
    if (!isEditing) throw new Error('Expected isEditing to be true.');
    if (wipDigits === null) throw new Error('Expected wipDigits to be non-null.');
    if (lightLength === null) throw new Error('Expected lightLength to be non-null.');

    if (ev.key === 'Backspace') {
      if (wipDigits.length <= 0) return;
      if (lightLength === 0) {
        // backspace when all digits are placeholder sets it to 00h00m00s
        setWipDigits([]);
      } else {
        const newDigits = [ ...wipDigits ];
        newDigits.pop();
        setWipDigits(newDigits);
        setLightLength(lightLength - 1);
      }
    } else if (ev.key === 'Enter') {
      saveAndQuit(true);
    } else if ('0' <= ev.key && ev.key <= '9') {
      const digit = Number(ev.key);
      if (0 > digit || digit > 9) throw new Error('Expected digit to be between 0 and 9 inclusive.');

      let initial: Array<number>;
      if (lightLength === 0) {
        initial = []; // discard the existing placeholder digits
      } else {
        initial = [ ...wipDigits ]; // we're in the process of typing digits
      }

      if (initial.length < 6) {
        setWipDigits([ ...initial, digit ]);
        setLightLength(lightLength + 1);
      }
    }
  };

  const children = (() => {
    let hhmmss: number[];

    if (isEditing) {
      if (wipDigits === null) throw new Error('Expected wipDigits to be non-null.');
      if (lightLength === null) throw new Error('Expected lightLength to be non-null.');

      hhmmss = padDigits(wipDigits);
    } else {
      // not editing
      if (wipDigits !== null) throw new Error('Expected wipDigits to be null.');
      if (lightLength !== null) throw new Error('Expected lightLength to be null.');

      hhmmss = trimDigits(digitsFromDuration(duration));
    }

    return renderDigits(hhmmss, lightLength);
  })();

  const innerClass = classNames("DurationInput-area", {
    editing: isEditing
  });

  return (
    <div
      className="DurationInput-main"
      onFocus={onFocus}
      onBlur={onBlur}
      onKeyDown={onKeyDown}
      tabIndex={0}
      role={isEditing ? "textbox" : "timer"}
    >
      <div className={innerClass}>
        {children}
      </div>
    </div>
  );
}

export default DurationInput;
export { padDigits, trimDigits };
