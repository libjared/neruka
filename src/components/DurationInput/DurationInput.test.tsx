import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import DurationInput, { padDigits, trimDigits } from '.';

const duration: Duration = { hours: 0, minutes: 15, seconds: 0 };
const noop = () => {};

it('matches snapshot', () => {
  const tree = renderer
    .create(<DurationInput duration={duration} onFinishEditing={noop} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('when the timer is focused', () => {
  it('switches to editing', () => {
    render(<DurationInput duration={duration} onFinishEditing={noop} />);
    expect(screen.queryByRole('textbox')).not.toBeInTheDocument();
    expect(screen.getByRole('timer')).toBeInTheDocument();
    fireEvent.focus(screen.getByRole('timer'));
    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(screen.queryByRole('timer')).not.toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<DurationInput duration={duration} onFinishEditing={noop} />);
    fireEvent.focus(screen.getByRole('timer'));
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('when the timer is blurred', () => {
  it('renders', () => {
    render(<DurationInput duration={duration} onFinishEditing={noop} />);
    fireEvent.focus(screen.getByRole('timer'));
    fireEvent.blur(screen.getByRole('textbox'));
  });

  xit('matches snapshot', () => {
    const { asFragment } = render(<DurationInput duration={duration} onFinishEditing={noop} />);
    fireEvent.focus(screen.getByRole('timer'));
    fireEvent.blur(screen.getByRole('textbox'));
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('when editing', () => {
  describe('when typing 4', () => {
    it('sets the timer text to 4 seconds', () => {
      render(<DurationInput duration={duration} onFinishEditing={noop} />);
      fireEvent.focus(screen.getByRole('timer'));
      fireEvent.keyDown(screen.getByRole('textbox'), { key: '4' });
      expect(screen.getByRole('textbox').textContent).toBe('00h00m04s');
    });

    xit('matches snapshot', () => {});
  });

  xit('matches snapshot', () => {});
});

describe('padDigits', () => {
  it('works', () => {
    expect(padDigits([1,5,0,0])).toEqual([0,0,1,5,0,0]);
  });
});

describe('trimDigits', () => {
  it('works', () => {
    expect(trimDigits([0,0,1,5,0,0])).toEqual([1,5,0,0]);
  });

  it('doesn\'t remove the last digit', () => {
    expect(trimDigits([0,0,0,0,0,0])).toEqual([0]);
  });
});
