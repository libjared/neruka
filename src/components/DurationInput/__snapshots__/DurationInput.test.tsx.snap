// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="DurationInput-main"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  role="timer"
  tabIndex={0}
>
  <div
    className="DurationInput-area"
  >
    <span
      className="Digit lit"
    >
      1
    </span>
    <span
      className="Digit lit"
    >
      5
    </span>
    <span
      className="Label minutes lit"
    >
      m
    </span>
    <span
      className="Digit lit"
    >
      0
    </span>
    <span
      className="Digit lit"
    >
      0
    </span>
    <span
      className="Label seconds lit"
    >
      s
    </span>
  </div>
</div>
`;

exports[`when the timer is focused matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="DurationInput-main"
    role="textbox"
    tabindex="0"
  >
    <div
      class="DurationInput-area editing"
    >
      <span
        class="Digit"
      >
        0
      </span>
      <span
        class="Digit"
      >
        0
      </span>
      <span
        class="Label hours"
      >
        h
      </span>
      <span
        class="Digit"
      >
        1
      </span>
      <span
        class="Digit"
      >
        5
      </span>
      <span
        class="Label minutes"
      >
        m
      </span>
      <span
        class="Digit"
      >
        0
      </span>
      <span
        class="Digit"
      >
        0
      </span>
      <span
        class="Label seconds"
      >
        s
      </span>
    </div>
  </div>
</DocumentFragment>
`;
